PluginFormcreatorForm::showForm() Use twig

btry opened the pull request:
use Twig in  PluginFormcreatorForm::showForm() 

Maybe merge this PR with all others twig conversions.

btry commented on front/form.form.php:
```diff
@@ -41,8 +41,12 @@
+         Html::redirect($computer->getLinkURL());
+      }
+   }
+   Html::back();
```
extra : better rediretion

btry commented on ajax/question_add.php:
```diff
@@ -44,7 +44,6 @@
     http_response_code(400);
     exit;
 }
-$question->getFromDB($question->getID()); // To remove if GLPI #9210 merged
```
the mentionned PR is merged

btry commented on ajax/question_delete.php:
```diff
@@ -39,6 +39,7 @@
 $questionId = (int) $_REQUEST['id'];
 $question = new PluginFormcreatorQuestion();
 if (!$question->getFromDB($questionId)) {
+    echo __('Question not found', 'formcreator');
```
if deletion fails, no error displayed, fixing it

btry replied:
The error mesage is processed by js/scripts.js. Your sgestion is possible, but also requires an update in the JS  code. This PR is complicated enough, then we may do this in a separate enhancement PR.

btry commented on css/styles.scss:
```diff
@@ -93,9 +93,15 @@ html, body {
+   // Fix ribbon position
+   position: relative;
+}
+
```
fix ribbon position when showForm() is used in a modal

btry commented on css/styles.scss:
```diff
@@ -93,9 +93,15 @@ html, body {
+
 /* Sections */
 #plugin_formcreator_form.plugin_formcreator_form_design {
-   > ol {
```
less strict CSS to match the expected elements under GLPI 10

btry commented on front/form.form.php:
```diff
@@ -41,8 +41,12 @@
+         Html::redirect($form->getLinkURL());
+      }
+   }
+   Html::back();
```
obey $_SESSION['glpibackcreated'] preference

btry commented on inc/common.class.php:
```diff
@@ -158,7 +158,9 @@ public static function prepareBooleanKeywords(string $keywords) : string {
-      $list = require_once(GLPI_PLUGIN_DOC_DIR . '/formcreator/' . self::getPictoFilename(GLPI_VERSION));
+      static $list = null;
+
+      $list = $list ?? require_once(GLPI_PLUGIN_DOC_DIR . '/formcreator/' . self::getPictoFilename());
```
avoid failure if the require statement is ran twice

btry commented on inc/common.class.php:
```diff
@@ -168,16 +170,16 @@ public static function getFontAwesomePictoNames() : array {
     * @param $version string GLPI version
     * @return string
     */
-   public static function getPictoFilename(string $version) : string {
```
obsolete argument removed

btry commented on inc/condition.class.php:
```diff
@@ -224,7 +226,7 @@ public function export(bool $remove_uuid = false) : array {
     * @return PluginFormcreatorCondition[]
     */
-   public function getConditionsFromItem(CommonDBTM $item) : array {
+   public static function getConditionsFromItem(CommonDBTM $item) : array {
```
make method callable from a template

btry commented on inc/condition.class.php:
```diff
@@ -249,6 +251,14 @@ public function getConditionsFromItem(CommonDBTM $item) : array {
       return $conditions;
    }
 
+   public static function countForItem(PluginFormcreatorConditionnableInterface $item): int {
```
make countElementsInTable callable from a template

btry commented on inc/condition.class.php:
```diff
@@ -276,153 +286,96 @@ public function showConditionsForItem(PluginFormcreatorConditionnableInterface $
-      // dropdown of questions
-      $form = PluginFormcreatorCommon::getForm();
-      $questionListExclusion = [];
+   public static function getQuestionsExclusion(PluginFormcreatorConditionnableInterface $item) {
```
isolate in dedicated method the code to exclude questions from a questions dropdown

btry commented on inc/field/actorfield.class.php:
```diff
@@ -249,11 +249,7 @@ protected function sanitizeValue($value): array {
-               } else {
-                  $knownUsers[$user->getID()] = $user->getRawName();
-               }
+               $knownUsers[$user->getID()] = $user->getFriendlyName();
```
remove backward compatibility with GLPI 9.5

btry commented on inc/fieldinterface.class.php:
```diff
@@ -291,4 +291,6 @@ public function isVisibleField() : bool;
     */
    public function getTranslatableStrings(array $options = []) : array;
+
+   public function getQuestion();
```
fix an IDE warning

AdrienClairembault commented:
If I remember correctly; this option was set to 'Yes' by default in the previous version:

![image](https://user-images.githubusercontent.com/42734840/148075911-109105a1-08bc-42c2-ba7c-dcd7ddc5df91.png)

-> This option was requested as an edge case, most people will want it set to 'Yes' for all their forms.

AdrienClairembault commented:
Visual issues when creating a section:

![image](https://user-images.githubusercontent.com/42734840/148077347-50d784f6-4802-42b4-b66a-f05abc29d906.png)

AdrienClairembault commented:
After adding my section I can't add any question, the tab has an error: 
![image](https://user-images.githubusercontent.com/42734840/148077536-b1395de1-5e7c-465b-82b9-542f4f20b21d.png)

AdrienClairembault commented:
Adding an actor to a target doesn't seem to work:

![image](https://user-images.githubusercontent.com/42734840/148078396-e4af3ed3-2cbc-4b68-8362-addf222a74f3.png)

After save:

![image](https://user-images.githubusercontent.com/42734840/148078442-6251a657-95b1-4a1c-a3db-a90ebeb7d49e.png)

AdrienClairembault commented:
I'm unable to edit the target conditions:

![image](https://user-images.githubusercontent.com/42734840/148078630-c17aa09b-9595-4eff-bfe3-7771c4a5fffc.png)

AdrienClairembault commented:
This buttons isn't doing anything:

![image](https://user-images.githubusercontent.com/42734840/148078858-db9dea33-df1a-42d8-92bc-9c9b100174d9.png)

I suppose it should toggle some search criteria ?

AdrienClairembault commented:
Visual issues when adding a new language to my form:

![image](https://user-images.githubusercontent.com/42734840/148079124-c4a3265b-87ea-41e4-a908-13baf563dd0c.png)

btry commented:
> Visual issues when trying to add a category from the "New item - Form" page:

this page is not yet converted into Twig. 

> Adding a validator to my form doens't seem to work

Fixed 

> Visual issues when creating a section (entity name + child entity checkbox):

Even when isEntityAssign returns false, the UI shows them. I tried to design templates as generic as possible, by using core's templates. The elements related to the entity are disabled, but I agree it would be better ti not show them. To do thos, it may be better to fix the template in the core.

About the empty ribbon : I did not focused on that for now. I find it not disturbing as is. Defining the icon for this itemtype should solve the topic. This is also true for question itemtype.

btry commented:
> Adding an actor to a target doesn't seem to work:

fixed

btry commented:
![image](https://user-images.githubusercontent.com/14139801/148235249-ed1a0c79-d57b-4d26-b31f-9f439bd407f7.png)

The toggle button is always displayed by GLPI. i think this page should not show the search filters, then GLPI should provide some mechanism to disable this toggle button.

btry commented:
ribbon in form language tab : 

needs 

```scss
.card_header {
   position:relative
}
```

This would also fix the other identified cases, but needs to be evaluated before inclusion in GLPI core.

AdrienClairembault commented:
> Even when isEntityAssign returns false, the UI shows them. I tried to design templates as generic as possible, by using core's templates. The elements related to the entity are disabled, but I agree it would be better ti not show them. To do thos, it may be better to fix the template in the core.

Done -> https://github.com/glpi-project/glpi/pull/10275.

AdrienClairembault reviewed (changes requested):
## UI Stuffs / Subjective

1\) Maybe we can add a few improvement to this page ?
![image](https://user-images.githubusercontent.com/42734840/149357111-bb144aa5-dcc4-4941-9237-c98715110b87.png)
![image](https://user-images.githubusercontent.com/42734840/149357165-1667bd83-1320-40f1-af6c-f62d27807899.png)

The input move when you switch between group and users, maybe a vertical layout would use the available space better (not very important).

2\) Maybe add some margin before displaying the first section:
![image](https://user-images.githubusercontent.com/42734840/149357787-2834589b-3e09-40c2-9209-8c6e06576ae8.png)
![image](https://user-images.githubusercontent.com/42734840/149357829-05017d13-95f7-4a51-bd2c-74a982b9c600.png)

Also the dark grey backgrounds seems a bit out of place in the new UI.
Maybe there is some tabler component we would use here but I'm not sure which.


## Bugs

1\) If I click on this dropdown, the modal close itself and my unsaved question is lost:
![image](https://user-images.githubusercontent.com/42734840/149359329-116c3dfb-a8b7-4b68-8cef-e8f19a688f3f.png)

Seems to behave differently depending on the question type.
Textarea closes the dropdown immediatly but some others don't.

2\) If I switch between all the possible values, new inputs keep popping up below the form:
![image](https://user-images.githubusercontent.com/42734840/149359852-fa92b931-0663-45ed-bd19-b0989774802e.png)

3\) If the section isn't saved yet there is never any input however:
![image](https://user-images.githubusercontent.com/42734840/149360009-e4d1ea85-7ff9-4736-b5c6-f6e213d3647a.png)

4\) Same as 2 but for the submit button conditions:
![image](https://user-images.githubusercontent.com/42734840/149360139-19ca9efa-aee5-4847-8549-6663e11ed5d9.png)

5\) Same as 3 but for a new question, can't add any condition:
![image](https://user-images.githubusercontent.com/42734840/149360277-6a9c8f5e-dbf2-45c7-a758-478b777f38f9.png)

6\) When creating a question, submit button doesn't close the modal.
Each time I press the button another question is added in the background.
Also can't close the modal.
![image](https://user-images.githubusercontent.com/42734840/149360666-2adff3fb-4738-421c-b6b0-be7afe3b9db6.png)

7\) Same as 2 but when editing a question:
![image](https://user-images.githubusercontent.com/42734840/149360915-05b9b36d-f23e-4c16-ada7-a6d1af18a45d.png)

8\) Not sure if caused by the twig changes but I can submit some partial values for targets (no question selected):
![image](https://user-images.githubusercontent.com/42734840/149361436-ff80786f-7c57-438f-b437-e79de5da40a4.png)

9\) Same as 2 but for target condition:
![image](https://user-images.githubusercontent.com/42734840/149361675-12f1a1f9-179a-4d1c-801c-11446862dafe.png)

10\) Can't add validator as a requester for a target change or problem:
![image](https://user-images.githubusercontent.com/42734840/149361963-6c93d271-065e-4de0-8a0f-b0af341367b9.png)
After submitting the value isn't added.
It does work for target ticket, I'm even able to add the user multiple time (is this ok ?):
![image](https://user-images.githubusercontent.com/42734840/149362172-58a51e9c-16ca-492c-a40d-71261dad4705.png)

11\) Actor questions aren't visible in the preview:
![image](https://user-images.githubusercontent.com/42734840/149362430-385635aa-88c9-411d-93b4-4c3fe4f908a0.png)
![image](https://user-images.githubusercontent.com/42734840/149362450-92bf2828-bee8-4a68-abc8-78b45bdbf369.png)

12\) Saving a question doesn't close the modal, can't close with the exit button either.

AdrienClairembault commented:
Another weird tab bug that I've been able to reproduce:
If you go to a form and:
- Go to "Targets" tab
- Go to "Access types" tabs
- Go back to "Targets" tab
- Add a new target
-> On submit, you are redirected to the "Access types" tab instead of staying on the "Targets" tab

btry commented:
I just noticed similar behaviours elsewhere. I'll fix all those things next week.

btry commented on RoboFile.php:
```diff
@@ -300,18 +300,7 @@ public function updateChangelog() {
-      $command.= " --keyword=_n:1,2,4t --keyword=__s:1,2t --keyword=__:1,2t --keyword=_e:1,2t --keyword=_x:1c,2,3t --keyword=_ex:1c,2,3t";
-      $command.= " --keyword=_sx:1c,2,3t --keyword=_nx:1c,2,3,5t";
-      $this->_exec($command);
+      $this->_exec('tools/extract_template.sh');
```
As Robo is being abandonned in GLPI project, I pick the locales extraction tool from GLPI, to find strings from twig templates

btry commented on ajax/question_add.php:
```diff
@@ -41,15 +41,15 @@
 
 if (!$question->add($_REQUEST)) {
-    http_response_code(400);
+    http_response_code(500);
```
err 500 is more appropriate here

btry commented on ajax/condition.php:
```diff
@@ -32,6 +32,31 @@
-echo $condition->getConditionHtml($_POST);
+$condition->fields['itemtype'] = $_POST['itemtype'];
+$condition->fields['items_id'] = (int) $_POST['items_id'];
+echo $condition->getConditionHtml($parent);
```
Atempt to simplify conditions creation and display for all conditionnable items. As conditions may be added to non-existing items (creating items and conditions at the same time), we need to forge items to make things work properly.

In other cases, loading them from DB is fine.

The most important need is to be able to find the form on top of the items, which are themselve parent of the conditions to show.

btry commented on ajax/question_add.php:
```diff
@@ -41,15 +41,15 @@
     'html'   => $question->getDesignHtml(),
 ];
-echo json_encode($json);
+echo json_encode($json, JSON_UNESCAPED_UNICODE);
```
usually, we shoud should prevent unicode chars escaping

btry commented on install/mysql/plugin_formcreator_empty.sql:
```diff
@@ -195,42 +195,42 @@ CREATE TABLE IF NOT EXISTS `glpi_plugin_formcreator_targetchanges` (
+  `ola_rule`                       int(11) NOT NULL DEFAULT '1',
+  `ola_question_tto`               int(11) NOT NULL DEFAULT '0',
+  `ola_question_ttr`               int(11) NOT NULL DEFAULT '0',
+  `uuid`                           varchar(255) DEFAULT NULL,
```
better indentation

btry commented on js/scripts.js:
```diff
@@ -464,8 +464,6 @@ var plugin_formcreator = new function() {
          },
          300);
-         // Remove empty rows
-         plugin_formcreator.moveUpItems(group);
```
the method is disabled, calling it is useless

btry commented on js/scripts.js:
```diff
@@ -548,7 +546,8 @@ var plugin_formcreator = new function() {
          return false;
-      });      var items = $(event.currentTarget).find('> .grid-stack-item');
+      });
+      var items = $(event.currentTarget).find('> .grid-stack-item');
```
style problem here

btry commented on js/scripts.js:
```diff
@@ -597,6 +596,7 @@ var plugin_formcreator = new function() {
          plugin_formcreator.dirty = false;
       }).done(function(response) {
          plugin_formcreator.dirty = false;
+         that.resetTabs();
```
when qhestions are moved, reset other tabs of the form to force reload of preview when the admin wants to test the form

btry commented on js/scripts.js:
```diff
@@ -624,8 +624,8 @@ var plugin_formcreator = new function() {
-      var item = $(target).closest('.grid-stack-item');
-      var id = item.attr('data-id');
+      var item = target.closest('.grid-stack-item');
+      var id = item.getAttribute('data-id');
```
switch to vanilla JS when possible

btry commented on js/scripts.js:
```diff
@@ -611,7 +611,7 @@ var plugin_formcreator = new function() {
             return;
          }
-         $(event.target).gridstack.update(
+         event.target.gridstack.update(
```
vanilla JS when possible

btry commented on js/scripts.js:
```diff
@@ -643,29 +643,10 @@ var plugin_formcreator = new function() {
-    * @param row  row to fill with items after it
-    */
-   this.moveUpItems = function (grid) {
-      return;
```
method was disabled (never  used in any release), dropping it

btry commented on js/scripts.js:
```diff
@@ -697,17 +678,18 @@ var plugin_formcreator = new function() {
 
    this.addQuestion = function () {
-      var form = $('form[data-itemtype="PluginFormcreatorQuestion"]');
+      var form = document.querySelector('form[data-itemtype="PluginFormcreatorQuestion"]');
```
vanilla JS if possible

btry commented on js/scripts.js:
```diff
@@ -697,17 +678,18 @@ var plugin_formcreator = new function() {
-         dataType: 'json'
+         data: new FormData(form),
+         processData: false,
+         contentType: false,
```
processData and contentType are here to help JQuery to work with FormData object; hels to remove useless JQuery use

btry commented on js/scripts.js:
```diff
@@ -697,17 +678,18 @@ var plugin_formcreator = new function() {
          displayAjaxMessageAfterRedirect();
       }).done(function(data) {
-         var sectionId = form.find('select[name="plugin_formcreator_sections_id"]').val();
+         var sectionId = form.querySelector('select[name="plugin_formcreator_sections_id"]').value;
```
less JQuery

btry commented on js/scripts.js:
```diff
@@ -726,30 +708,29 @@ var plugin_formcreator = new function() {
-         type: "POST",
          data: form.serializeArray(),
-         dataType: 'html'
+         dataType: 'json'
```
returned body switched to JSON

btry commented on js/scripts.js:
```diff
@@ -726,30 +708,29 @@ var plugin_formcreator = new function() {
       }).done(function(data) {
          var question = $('.plugin_formcreator_form_design[data-itemtype="PluginFormcreatorForm"] [data-itemtype="PluginFormcreatorQuestion"][data-id="' + questionId + '"]');
-         question.find('[data-field="name"]').text(data)
+         question.find('[data-field="name"]').text(data['name'])
```
returned body switched to JSON

btry commented on js/scripts.js:
```diff
@@ -726,30 +708,29 @@ var plugin_formcreator = new function() {
-      var item = $(target).closest('.grid-stack-item');
-      var id = item.attr('data-id');
+      var item = target.closest('.grid-stack-item');
+      var id = item.getAttribute('data-id');
```
less JQuery

btry commented on js/scripts.js:
```diff
@@ -888,13 +869,28 @@ var plugin_formcreator = new function() {
+               }
+               $('#' + modalId).modal('hide');
+            });
+         }
```
fix non closing modals.

btry commented on js/scripts.js:
```diff
@@ -922,52 +918,76 @@ var plugin_formcreator = new function() {
+               }
+               $('#' + modalId).modal('hide');
+            });
+         }
```
fix non closing modal

btry commented on js/scripts.js:
```diff
@@ -922,52 +918,76 @@ var plugin_formcreator = new function() {
+         processData: false,
+         contentType: false,
+         data: new FormData(form),
+         dataType: 'json'
```
switch to json response

btry commented on js/scripts.js:
```diff
@@ -922,52 +918,76 @@ var plugin_formcreator = new function() {
 
-   this.addSection = function () {
-      var form = $('form[data-itemtype="PluginFormcreatorSection"]');
+   this.addSection = function (event) {
```
event pssed to function, helping to mot search too much in the DOM

btry commented on js/scripts.js:
```diff
@@ -922,52 +918,76 @@ var plugin_formcreator = new function() {
-   this.editSection = function () {
-      var form = $('form[data-itemtype="PluginFormcreatorSection"]');
-      var sectionId = form.find('[name="id"]').val();
+   this.editSection = function (event) {
```
event passed to function, here as well

btry commented on js/scripts.js:
```diff
@@ -1122,6 +1148,31 @@ var plugin_formcreator = new function() {
       });
    }
 
+   this.changeActor = function(type, value) {
```
factorization of :
plugin_formcreator_ChangeActorRequester
plugin_formcreator_ChangeActorWatcher
plugin_formcreator_ChangeActorASsigned

btry commented on js/scripts.js:
```diff
@@ -1313,81 +1388,6 @@ function plugin_formcreator_hideAssignedForm() {
    $('#btn_cancel_assigned').hide();
 }
 
-function plugin_formcreator_ChangeActorRequester(value) {
```
factorized

btry commented on js/scripts.js:
```diff
@@ -1313,81 +1388,6 @@ function plugin_formcreator_hideAssignedForm() {
-   }
-}
-
-function plugin_formcreator_ChangeActorWatcher(value) {
```
factorized

btry commented on js/scripts.js:
```diff
@@ -1313,81 +1388,6 @@ function plugin_formcreator_hideAssignedForm() {
-   }
-}
-
-function plugin_formcreator_ChangeActorAssigned(value) {
```
factorized

btry commented on js/scripts.js:
```diff
@@ -1169,6 +1220,41 @@ var plugin_formcreator = new function() {
       );
    }
+
+   this.deleteActor = function (item) {
```
convert add / delete of target actors calls into Ajax, for better UX.

btry commented on js/scripts.js:
```diff
@@ -1183,17 +1269,6 @@ function plugin_formcreator_addTarget(items_id) {
-         id: items_id
-      },
-   });
-}
```
no longer used

btry commented on js/scripts.js:
```diff
@@ -1517,32 +1517,26 @@ function plugin_formcreator_toggleCondition(target) {
 
 function plugin_formcreator_removeNextCondition(target) {
-   $(target).parents('tr').remove();
+   target.closest('div.row').remove();
```
update to new DOM, drop jquery usage

btry commented on tests/3-unit/PluginFormcreatorCondition.php:
```diff
@@ -35,7 +35,7 @@ class PluginFormcreatorCondition extends CommonTestCase {
       parent::beforeTestMethod($method);
 
-      self::login('glpi', 'glpi');
+      $this->login('glpi', 'glpi');
```
the methid is not static

btry replied:
the methid is not static

btry commented on tests/3-unit/PluginFormcreatorCondition.php:
```diff
@@ -82,14 +82,14 @@ public function testGetConditionsFromItem() {
       $this->boolean($condition->isNewItem())->isFalse();
 
       // Check that all conditions are retrieved
-      $output = $condition->getConditionsFromItem($question);
```
method is now static

btry replied:
method is now static

btry commented on tests/3-unit/PluginFormcreatorCondition.php:
```diff
@@ -82,14 +82,14 @@ public function testGetConditionsFromItem() {
       $question = $this->getQuestion();
       $form = new \PluginFormcreatorForm();
-      $form->getFromDBByQuestion($question);
+      $form = \PluginFormcreatorForm::getByItem($question);
```
use a more generic method to find a form from an item inside it

btry commented on tests/3-unit/PluginFormcreatorCondition.php:
```diff
@@ -144,11 +144,9 @@ public function testExport() {
       $question2 = $this->getQuestion([
          'plugin_formcreator_forms_id' => $form->getID(),
-      ]);
-      $question2->updateConditions([
```
adding conditions can be done while adding the question

btry commented on inc/form.class.php:
```diff
@@ -2198,52 +1956,57 @@ public function isPublicAccess() : bool {
+    * @param CommonDBTM $item
+    * @return null|self
     */
-   public function getFromDBBySection(PluginFormcreatorSection $item) {
```
replaced by more generic getByItem method

btry commented on inc/form.class.php:
```diff
@@ -2198,52 +1956,57 @@ public function isPublicAccess() : bool {
-      return $this->getFromDB($item->fields[self::getForeignKeyField()]);
-   }
 
-   public function getFromDBByQuestion(PluginFormcreatorQuestion $question) {
```
replaced by more generic getByItem method

btry commented on inc/question.class.php:
```diff
@@ -755,6 +739,14 @@ public function post_purgeItem() {
+      //    'item' => $this,
+      //    'params' => $options,
+      // ]);
+      // return true;
```
too much complex to convert now, lets keep this draft and continue in an other PR

btry commented on inc/section.class.php:
```diff
@@ -86,6 +101,12 @@ public function rawSearchOptions() {
+      return [
+         'update', 'clone', 'add_note',
+      ];
+   }
```
forbid some massive actions

btry commented on inc/target_actor.class.php:
```diff
@@ -92,6 +92,39 @@ public static function getTypeName($nb = 0) {
    }
 
    public function prepareInputForAdd($input) {
+      $requiredKeys = ['itemtype', 'items_id', 'actor_role', 'actor_type'];
```
add integrity checks, does not handles all cases for now, but fix the inconsistency found this week

btry commented on inc/targetproblem.class.php:
```diff
@@ -669,6 +558,58 @@ public static function countItemsToImport(array $input) : int {
       return 1 + self::countChildren($subItems, $input);
    }
 
+   public function defineTabs($options = []) {
```
like target ticket and target change : swtch to a multiple tabs design to edit this target

btry commented on inc/targetticket.class.php:
```diff
@@ -794,20 +697,6 @@ public function pre_purgeItem() {
-      if ($this->input['show_rule'] != PluginFormcreatorCondition::SHOW_RULE_ALWAYS) {
-         $this->updateConditions($this->input);
-      }
-   }
```
factorize with all targets types, see implementation in AbstractTarget. 

If adding actors by default is not wanted, then create for each class a method to add default actors.

btry commented on inc/targetticket.class.php:
```diff
@@ -1144,7 +1033,7 @@ protected  function showTypeSettings($rand) {
       echo '</div>';
       echo '<div id="requesttype_question_value" style="display: none">';
       PluginFormcreatorQuestion::dropdownForForm(
```
dropdownForForm now gets form object as 1st arg

AdrienClairembault reviewed (changes requested):
1\) The visibility condition dropdown is too large, it doesn't seem to fit the input:
![image](https://user-images.githubusercontent.com/42734840/150139195-e658bcc7-c1e2-41cb-9ba9-0c8105c478ba.png)

Also, if you have a scroll-able page (a big question form like the checkbox question type for example), the options are disconnected from the input:

![image](https://user-images.githubusercontent.com/42734840/150139413-2828b899-aaa9-4e1a-9c0c-58392b2ce1f6.png)

2\) When creating a form, "All langages" is displayed differently when selected vs the value in the dropdown:
![image](https://user-images.githubusercontent.com/42734840/150139823-baa72483-95af-4c8d-bd3b-4f9b58fd15f5.png)
  
Also it should be  "All lang**u**ages" instead of  "All langages".

3\) I'm not sure what happened but the first form I tried to create didn't had the correct default values.
The visible option was on "No", the default language on "All Languages" and both colors to white #FFFFFF.
After submitting the form it was also not saved in the database instead it was used to modify an existing form:
![image](https://user-images.githubusercontent.com/42734840/150142433-415230ac-c021-40e2-87c6-3623f3c7e1dc.png)
I'm 100% sure of it as it was the exact colors and icons I had picked + the correct time.

Every other form I tried to create after this didn't have the issue so I wasn't able to reproduce this bug.

UPDATE, I've been able to reproduce it: 
- Save the questions on an existing form
- Try to create a new form

![image](https://user-images.githubusercontent.com/42734840/150143434-1f2f2f6d-1386-4d2d-b09e-d156ed129b66.png)

4\) The validation input still "moves around" depending on the selected value
![image](https://user-images.githubusercontent.com/42734840/150140483-114d36fc-355c-4e27-87bc-dfa3ad1f36b2.png)
It should use the new bootstrap col/row forms instead of the legacy html table.

The label "Need validaton?" has a spelling mistake.


5\) If I click twice on the "move section down" button (which I first though was a button to hide the section content, not move my section. not sure how to improve this UI wise) then some questions are not shown anymore.

Before:
![image](https://user-images.githubusercontent.com/42734840/150144105-d5ee8943-a6f7-4c5c-8a66-a4925c4bb02e.png)

After: 
![image](https://user-images.githubusercontent.com/42734840/150144162-514d910f-dbc5-40c7-bc14-d5cc346fabfa.png)


6\) The submit button visibility condition trigger an error:
![image](https://user-images.githubusercontent.com/42734840/150145175-4a46c502-1c37-4042-bc99-dbdaa31b43c4.png)

You need to set a condition and save, the error wont be shown if no conditions are set.

Seems to be OK for questions and sections.

7\) I can't update my target ticket, right errors.
![image](https://user-images.githubusercontent.com/42734840/150145942-cbc8b4eb-fa76-4a4f-9eec-276b3887887d.png)

This is a GLPI with only formcreator and no rights / configuration modification.

 8\) Two issues with the actor fields:
![image](https://user-images.githubusercontent.com/42734840/150146308-1e1bf612-d9e0-4e52-ad06-365436be9156.png)

There is an excessive margin compared to similar inputs used by GLPI:
![image](https://user-images.githubusercontent.com/42734840/150146570-ac581618-3221-446a-9528-b1e73d6eadf9.png)

Also the default "----" option shouldn't be selectable (might be fixed by https://github.com/glpi-project/glpi/pull/10288)

9\) I can't click on these icons in a form target, seems to do nothing:
![image](https://user-images.githubusercontent.com/42734840/150147096-e0fc7aed-4818-4f31-a505-00446d43f7aa.png)

Beside not having a "pointer" cursor, the 'delete' button seems to trigger a 404 error:
![image](https://user-images.githubusercontent.com/42734840/150147200-4df4a361-352d-43fe-8fab-93e1c7dd2dda.png)
The "notification" button doesn't seem to do anything.

Same 404 error if I try to add another actor.

10\) The "issues" button lead me to an error:

![image](https://user-images.githubusercontent.com/42734840/150147683-f2eeb587-3b90-4455-95bf-840b139c296f.png)
![image](https://user-images.githubusercontent.com/42734840/150147651-e551ccca-415c-4511-8828-c6964b824f90.png)

The issue seems to be that I'm able to add a "Tag" question to my forms without having the tag plugin installed or activated.
This also prevent me from displaying this form.

I think firstly I shouldn't be able to add a Tag question if I don't have the plugin activated.
Even tho, this kind of case might still happen (If I enable tag, create a tag question then disable the tag plugin for example) so we should avoid throwing an error and just skip the question.

11\) No breadcrumbs when searching or displaying form answers:
![image](https://user-images.githubusercontent.com/42734840/150148116-a1179328-7d9b-430b-aceb-44bc67986b61.png)


12\) Error on target change:
![image](https://user-images.githubusercontent.com/42734840/150149054-5fff2b9c-bc6f-4531-85b6-1fb3899809ff.png)

13\) I have a test form with 6 targets, when I submit it 0 of them are created.

Here is the form export (had to save it as .json.txt as json file are forbidden it seems).

[export_formcreator_20220119_1521.json.txt](https://github.com/pluginsGLPI/formcreator/files/7897870/export_formcreator_20220119_1521.json.txt)

14\) I have this warning in my logs:
```
[2022-01-19 15:01:53] glpiphplog.NOTICE:   *** PHP Notice (8): Undefined index: is_captcha_enabled in /home/adrie/localhost/gmaster2/plugins/formcreator/front/form_profile.form.php at line 48  
```

btry commented:
> 1) The visibility condition dropdown is too large, it doesn't seem to fit the input:

Reduced; but the width is not consistent everywhere. There is a problem with design of questions. This form is not converted into Twig and I integrated the twig version of conditions rendering in the old code. 

A full technical redesign of this part is needed because of several reasons (incoming devs, a communautary problem report, migration to twig, constant addition of options making this part had to maintain, bug prone). 

I want to fix that in a distinct PR to have a chance to submit a readable and understandable code design. This PR is too big and spreads on to many topics (fixes unrelated ti twig conversion) now.

> Also, if you have a scroll-able page (a big question form like the checkbox question type for example), the options are disconnected from the input:

should  be fixed with a speecific PR, see above comment

> 2) When creating a form, "All langages" is displayed differently when selected vs the value in the dropdown

fixed in GLPI, was a RTL problem. Typo fixed.

> 3) I'm not sure what happened but the first form I tried to create didn't had the correct default values.

fixed. Was a problem of data stored in session, not erased when a updating form properties, not changing the form object itself. see PluginFormcreatorForm::post_updateItem:() 

Maybe a better solution should be found as it is currently hacky.


> 4) The validation input still "moves around" depending on the selected value

out of topic of this PR : this tab is not converted into Twig yet. Also not very harmful, other things have priority

AdrienClairembault reviewed (approved):
OK for now, it's hard to keep track of everything with so many comments, commits and changes.
I'll test everything again later (next beta release maybe ?)

btry commented:
I'm still fixing last issues; i'll update the previous comment. I('ll tell you when tie PR is ready for a new review
